lib/ABI/AddABIAwarenessToFunctionPrototypes.cpp

```diff
@@ -0,0 +1,210 @@
+      Result = Result && rc_recur needsVectorRegisterSupportImpl(VH, F.Type());
+    break;
+
+  case model::TypeKind::CABIFunctionType: {
+    Result = true;
+    using CABIFT = model::CABIFunctionType;
+    for (const auto &A : llvm::cast<CABIFT>(&Type)->Arguments())
+      Result = Result && rc_recur needsVectorRegisterSupportImpl(VH, A.Type());
+    const auto &ReturnType = llvm::cast<CABIFT>(&Type)->ReturnType();
+    Result = Result && rc_recur needsVectorRegisterSupportImpl(VH, ReturnType);
+  } break;
+
+  case model::TypeKind::RawFunctionType: {
+    Result = true;
+    using RawFT = model::RawFunctionType;
+    for (const auto &A : llvm::cast<RawFT>(&Type)->Arguments()) {
+      auto Kind = model::Register::primitiveKind(A.Location());
+      Result = Result && rc_recur needsVectorRegisterSupportImpl(VH, Kind);
+      Result = Result && rc_recur needsVectorRegisterSupportImpl(VH, A.Type());
+    }
+    for (const auto &V : llvm::cast<RawFT>(&Type)->ReturnValues()) {
+      auto Kind = model::Register::primitiveKind(V.Location());
+      Result = Result && rc_recur needsVectorRegisterSupportImpl(VH, Kind);
+      Result = Result && rc_recur needsVectorRegisterSupportImpl(VH, V.Type());
+    }
+
+    const auto &Stack = llvm::cast<RawFT>(&Type)->StackArgumentsType();
+    revng_assert(Stack.Qualifiers().empty());
+    const model::TypePath &Internal = Stack.UnqualifiedType();
+    if (Internal.isValid())
+      Result = Result && rc_recur needsVectorRegisterSupportImpl(VH, Internal);
+  } break;
+
+  default:
+    revng_abort("Unknown type.");
+  }
+
+  if (Result)
+    VH.setVerified(&Type);
+
+  VH.verificationCompleted(&Type);
+
+  rc_return VH.maybeFail(Result);
+}
+
+static bool needsVectorRegisterSupport(model::VerifyHelper &VH,
+                                       const model::TypePath &Type) {
+  return !needsVectorRegisterSupportImpl(VH, model::QualifiedType(Type, {}));
+}
+
+using namespace std::string_literals;
+
+class AddABIAwarenessToFunctionPrototypes {
+public:
+  static constexpr auto Name = "AddABIAwarenessToFunctionPrototypes";
+  inline static const std::tuple Options = { pipeline::Option("abi",
+                                                              "Invalid"s) };
+  std::vector<std::vector<pipeline::Kind *>> AcceptedKinds = {};
+
+  void run(pipeline::Context &Context, std::string InputABI) {
+    auto &Model = revng::getWritableModelFromContext(Context);
+
+    // Minimize the negative impact on binaries with ABI that is not fully
+    // supported by disabling the conversion by default.
+    //
+    // Use revng model pass on those functions to convert them manually.
+    //
+    // TODO: extend this list.
+    static constexpr std::array ABIsTheConversionIsEnabledFor = {
+      model::ABI::SystemV_x86_64,        model::ABI::SystemV_x86,
+      model::ABI::SystemV_x86_regparm_3, model::ABI::SystemV_x86_regparm_2,
+      model::ABI::SystemV_x86_regparm_1, model::ABI::AAPCS
```

Comment:
* Is `clang-format` leaveing all those spaces?
* Why do we limit ourselves to these ABIs? What's the point of having all the others?

Reply:
1. It's worse, `clang-format` is the one who inserted them in there. I'd rather they be listed one element per line.
2. These are the ones that are currently tested. The others have problems with them that need a bit more looking into before we can "use" them with any kind of certainty.

Reply:
2. Mmmh. I see a great opportunity of forgetting about these. I'd leave them all enabled or disabled them entirely and add a comment to the top of the YAML.

Reply:
I don't think I get what you mean.

My current stance is that the `ConvertToCABIFunctionType` *analysis* only works on the selected subset of ABIs that are tested (which will probably get wider soon) - it does nothing for any function that doesn't apply, while `tryConvertToCABI` c++ *function* does best effort conversion on anything you give it (BUT to access it you need to do so from c++ code, so it's not user-facing) with the downside of possibly breaking everything.

Reply:
`tryConvertToCABI` is and will only used by `ConvertToCABIFunctionType`, so it's like it doesn't exist.
Either we disable it entirely or expose them to the user. If we keep this, we'll have a tier 1 and 2 calling convention for a long time, I don't really want this. I prefer to get crashes and bug reports as an incentive to fix things.

Reply:
For once, I already know the cases where all the position based ABI (Microsoft x64 and SystemZ) and AArch64 (the problem is similar, so it needs to be handled together with position based ones) don't work, so we'll probably run into problems if we try to use them on a real binary. I wouldn't want to expose the other developers (mainly Pietro) to those problems and make them waste time debugging something that just isn't ready yet.

As in, I wouldn't be confident putting the analysis into the auto-analysis pipeline without this check: we're introducing too many unnecessary bugs in a place that ever user uses by default.

P. S. As for mips, I have no idea what state it's in, I never got around to even trying to test it 😅, I had my hands occupied even without it.

Reply:
OK, then let's not register them altogether. If you want to test them you can re-enable them.

Reply:
I'm still somewhat confused about what you mean by not registering them. Removing them from `model::ABI::Values`? That sounds a bit too harsh if you ask me. Asserting when you try to `abi::Definition::get`? Even worse.

Reply:
OK, as you prefer.
Please put in your pad that this needs testing+fixing though.

Reply:
I'm still a bit confused about what's the consensus then?

What should I do (on top of updating the pad)? Should I keep this list here? Move it to `tryConvertToCABI`? Remove it?

Reply:
I tried to persuade you to drop it, but it's "too harsh" `:)`, so as you wish, we can leave this here.